### PR TITLE
🌈 feat: StatelessCNI: Applying stateless CNI mode changes in network package.

### PR DESCRIPTION
**Pull request by behzad-mir**
**Reason for Change**:
Applying the required stateless CNI mode changes in network package.

**Issue Fixed**:

**Requirements**:

- [X] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [X] relevant PR labels added

**Notes**:

### CODE REVIEW

File: network/manager.go [hunk 1]
```diff
@@ -354,11 +403,56 @@ func (nm *networkManager) CreateEndpoint(cli apipaClient, networkID string, epIn
...
 }
 
+// UpdateEndpointState will make a call to CNS updatEndpointState API in the stateless CNI mode
+// It will add HNSEndpointID or HostVeth name to the endpoint state
+func (nm *networkManager) UpdateEndpointState(ep *endpoint) error {
+	cnilogger.Logger.Info("Calling cns updateEndpoint API with ", zap.String("containerID: ", ep.ContainerID), zap.String("HnsId: ", ep.HnsId), zap.String("HostIfName: ", ep.HostIfName))
+	response, err := nm.CnsClient.UpdateEndpoint(context.TODO(), ep.ContainerID, ep.HnsId, ep.HostIfName)
+	if err != nil {
+		cnilogger.Logger.Error("Update endpoint API returend with error", zap.Error(err))
+		return errors.Wrapf(err, "Get endpoint API returend with error")
```
Comment (vipul-21) [inline network/manager.go:413]: The error description is saying `Get endpoint` instead of `Update endpoint`

File: network/manager.go [hunk 2]
```diff
@@ -354,11 +403,56 @@ func (nm *networkManager) CreateEndpoint(cli apipaClient, networkID string, epIn
...
+
 // DeleteEndpoint deletes an existing container endpoint.
-func (nm *networkManager) DeleteEndpoint(networkID, endpointID string) error {
+func (nm *networkManager) DeleteEndpoint(networkID, endpointID string, epInfo *EndpointInfo) error {
 	nm.Lock()
 	defer nm.Unlock()
 
+	if nm.IsStatelessCNIMode() {
+		nw := &network{
+			Id:           "azure",
```
Comment (vipul-21) [inline network/manager.go:426]: We can use the `networkID` here ?

File: network/manager.go [hunk 3]
```diff
@@ -382,6 +476,31 @@ func (nm *networkManager) GetEndpointInfo(networkId string, endpointId string) (
 	nm.Lock()
 	defer nm.Unlock()
 
+	if nm.IsStatelessCNIMode() {
+		cnilogger.Logger.Info("calling cns getEndpoint API")
+		endpointResponse, err := nm.CnsClient.GetEndpoint(context.TODO(), endpointId)
+		if err != nil {
+			cnilogger.Logger.Error("Get endpoint API returend with error", zap.Error(err))
+			return nil, errors.Wrapf(err, "Get endpoint API returend with error")
```
Comment (behzad-mir) [inline network/manager.go:484]: @tamilmani1989 
I can check in here if the state is not found we return err=nil. But I'm afarid we might create other issues and prevent a cleanup.

File: network/manager.go [hunk 4]
```diff
@@ -298,6 +340,7 @@ func (nm *networkManager) DeleteNetwork(networkID string) error {
 
 // GetNetworkInfo returns information about the given network.
 func (nm *networkManager) GetNetworkInfo(networkId string) (NetworkInfo, error) {
+
```
Comment (tamilmani1989) [inline network/manager.go:343]: nit: space.. revert this change

File: network/manager.go [hunk 5]
```diff
@@ -326,6 +369,7 @@ func (nm *networkManager) GetNetworkInfo(networkId string) (NetworkInfo, error)
 
 // CreateEndpoint creates a new container endpoint.
 func (nm *networkManager) CreateEndpoint(cli apipaClient, networkID string, epInfo *EndpointInfo) error {
+
```
Comment (tamilmani1989) [inline network/manager.go:372]: nit: space.. revert

File: network/manager.go [hunk 6]
```diff
@@ -336,16 +380,21 @@ func (nm *networkManager) CreateEndpoint(cli apipaClient, networkID string, epIn
 
 	if nw.VlanId != 0 {
 		if epInfo.Data[VlanIDKey] == nil {
+			cnilogger.Logger.Info("overriding endpoint vlanid with network vlanid")
```
Comment (tamilmani1989) [inline network/manager.go:383]: why logging again? can you remove this

File: network/manager.go [hunk 7]
```diff
@@ -269,7 +307,9 @@ func (nm *networkManager) CreateNetwork(nwInfo *NetworkInfo) error {
 	if err != nil {
 		return err
 	}
-
+	if nm.IsStatelessCNIMode() {
```
Comment (tamilmani1989) [inline network/manager.go:310]: can we move this check inside nm.save() instead of repeating in multiple places?

File: network/manager.go (hunk 2 above)
Comment (tamilmani1989) [inline network/manager.go:426]: lets not consider `azure` here instead lets write api to get network name.. incase of multitenancy network name has to be derived from vlan id and subnet space

File: network/manager.go [hunk 8]
```diff
@@ -354,11 +403,56 @@ func (nm *networkManager) CreateEndpoint(cli apipaClient, networkID string, epIn
...
 	nm.Lock()
 	defer nm.Unlock()
 
+	if nm.IsStatelessCNIMode() {
+		nw := &network{
+			Id:           "azure",
+			Mode:         opModeTransparentVlan,
+			SnatBridgeIP: "",
+			extIf: &externalInterface{
+				Name:       "eth0",
```
Comment (tamilmani1989) [inline network/manager.go:430]: lets not hardcode interface name as well

File: network/manager.go [hunk 9]
```diff
@@ -354,11 +403,56 @@ func (nm *networkManager) CreateEndpoint(cli apipaClient, networkID string, epIn
...
+				MacAddress: nil,
+			},
+		}
+
```
Comment (tamilmani1989) [inline network/manager.go:434]: i would suggest to move this block to separate function

File: network/manager.go [hunk 10]
```diff
@@ -354,11 +403,56 @@ func (nm *networkManager) CreateEndpoint(cli apipaClient, networkID string, epIn
...
+			HostIfName:               epInfo.IfName,
+			LocalIP:                  "",
+			VlanID:                   0,
+			AllowInboundFromHostToNC: false,
+			AllowInboundFromNCToHost: false,
+			EnableSnatOnHost:         false,
+			EnableMultitenancy:       false,
+			NetworkContainerID:       epInfo.Id,
+		}
+		netlink.NewNetlink()
```
Comment (tamilmani1989) [inline network/manager.go:447]: why this is required?

File: network/manager.go [hunk 11]
```diff
@@ -354,11 +403,56 @@ func (nm *networkManager) CreateEndpoint(cli apipaClient, networkID string, epIn
...
+		cnilogger.Logger.Info("Deleting endpoint with", zap.String("Endpoint Info: ", epInfo.PrettyString()), zap.String("HNISID : ", ep.HnsId))
+		err := nw.deleteEndpointImpl(netlink.NewNetlink(), platform.NewExecClient(), nil, ep)
+		if err != nil {
+			return err
+		}
+		return nil
```
Comment (tamilmani1989) [inline network/manager.go:453]: can it be return err?

File: network/manager.go [hunk 12]
```diff
@@ -382,6 +476,31 @@ func (nm *networkManager) GetEndpointInfo(networkId string, endpointId string) (
...
+	if nm.IsStatelessCNIMode() {
+		cnilogger.Logger.Info("calling cns getEndpoint API")
+		endpointResponse, err := nm.CnsClient.GetEndpoint(context.TODO(), endpointId)
+		if err != nil {
+			cnilogger.Logger.Error("Get endpoint API returend with error", zap.Error(err))
+			return nil, errors.Wrapf(err, "Get endpoint API returend with error")
+		}
+		epInfo := &EndpointInfo{
+			Id:                 endpointId,
+			IfIndex:            0, // Azure CNI supports only one interface
```
Comment (tamilmani1989) [inline network/manager.go:488]: For ifindex, we should  return whatever existing cni returns.. if required we can get interface index based on interface name

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: network/manager.go (hunk 3 above)
Comment (tamilmani1989) [reply network/manager.go:484]: should check if error is `endpoint not found` and then ignore.. cannot ignore all other errors

File: cni/network/network.go [hunk 13]
```diff
@@ -1052,11 +1056,13 @@ func (plugin *NetPlugin) Delete(args *cniSkel.CmdArgs) error {
 				// if cni hits this, mostly state file would be missing and it can be reboot scenario where
 				// container runtime tries to delete and create pods which existed before reboot.
 				err = nil
-				return err
+				if !plugin.nm.IsStatelessCNIMode() {
```
Comment (vipul-21) [inline cni/network/network.go:1059]: nit: Can we add a comment here why we need to do this for stateless CNI ?

File: network/manager.go [hunk 14]
```diff
@@ -113,6 +125,10 @@ func NewNetworkManager(nl netlink.NetlinkInterface, plc platform.ExecClient, net
 func (nm *networkManager) Initialize(config *common.PluginConfig, isRehydrationRequired bool) error {
 	nm.Version = config.Version
 	nm.store = config.Store
+	if config.Stateless == true {
+		nm.SetStatelessCNIMode()
```
Comment (vipul-21) [inline network/manager.go:129]: We are bypassing the error returned by `SetStatelessCNIMode`, we should return that error

File: network/manager.go [hunk 15]
```diff
@@ -61,6 +69,8 @@ type EndpointClient interface {
 
 // NetworkManager manages the set of container networking resources.
 type networkManager struct {
+	StetlessCniMode    bool
```
Comment (vipul-21) [inline network/manager.go:72]:
```suggestion
	StatelessCniMode    bool
```

File: network/manager.go [hunk 16]
```diff
@@ -123,6 +139,26 @@ func (nm *networkManager) Initialize(config *common.PluginConfig, isRehydrationR
 func (nm *networkManager) Uninitialize() {
 }
 
+// SetStatelessCNIMode enable the statelessCNI falg and inititlizes a CNSClient
+func (nm *networkManager) SetStatelessCNIMode() error {
+	nm.StetlessCniMode = true
```
Comment (vipul-21) [inline network/manager.go:144]:
```suggestion
	nm.StatelessCniMode = true
```

File: network/manager.go [hunk 17]
```diff
@@ -123,6 +139,26 @@ func (nm *networkManager) Initialize(config *common.PluginConfig, isRehydrationR
...
+	if err != nil {
+		return errors.Wrapf(err, "failed to initialize CNS client")
+	}
+	nm.CnsClient = client
+	return nil
+}
+
+// IsStatelessCNIMode checks if the Stateless CNI mode has been enabled or not
+func (nm *networkManager) IsStatelessCNIMode() bool {
+	if nm.StetlessCniMode == true {
```
Comment (vipul-21) [inline network/manager.go:156]:
```suggestion
	if nm.StatelessCniMode == true {
```

File: cni/network/network.go [hunk 18]
```diff
@@ -150,16 +150,19 @@ func (plugin *NetPlugin) Start(config *common.PluginConfig) error {
 	// Log platform information.
 	log.Logger.Info("Plugin Info",
 		zap.String("name", plugin.Name),
-		zap.String("version", plugin.Version))
+		zap.String("version", plugin.Version),
+		zap.String("component", "cni-net"))
```
[the remark this answers is not in the document]
Comment (behzad-mir) [reply cni/network/network.go:154]: addressed in the next commit

File: cni/network/network.go (hunk 13 above)
Comment (behzad-mir) [reply cni/network/network.go:1059]: addressed in the next commit

File: network/manager.go (hunk 14 above)
Comment (behzad-mir) [reply network/manager.go:129]: addressed in the next commit.

File: network/manager.go (hunk 7 above)
Comment (behzad-mir) [reply network/manager.go:310]: addressed in the next commit.

File: network/manager.go (hunk 1 above)
Comment (behzad-mir) [reply network/manager.go:413]: addressed in the next commit

File: network/manager.go (hunk 2 above)
Comment (behzad-mir) [reply network/manager.go:426]: addressed in the next commit

Comment (behzad-mir) [reply network/manager.go:426]: I am using the networkId that is passed tot he func for now.

File: network/manager.go (hunk 9 above)
Comment (behzad-mir) [reply network/manager.go:434]: addressed in the next commit

File: network/manager.go (hunk 10 above)
Comment (behzad-mir) [reply network/manager.go:447]: forgot to clean up.

File: network/manager.go (hunk 3 above)
Comment (behzad-mir) [reply network/manager.go:484]: We should still return err here back to the cni/network.go Delete func so it can just exit gracefully the same as legacy CNI.

File: network/manager.go (hunk 12 above)
Comment (behzad-mir) [reply network/manager.go:479]: This value Has been hardcoded on existing GetEndpointInfo API with 0. I just followed the same. For the new commit, I just added a constant value.

File: network/manager.go (hunk 8 above)
Comment (behzad-mir) [reply network/manager.go]: A constant has been added.
